### PR TITLE
Add DenseNet models and experiments

### PR DESCRIPTION
**Pull request by citerana**
This PR adds the densenet121 and densenet169 models to the project. Both models support pretraining with imagenet weights. Additionally, this includes a series of experiments that were used to investigate the relative performances of the different architectures with variable learning rate schedules.

The link to the imagenet weights can be found at the top level comment of each model. These models use a tensorflow backend so choose the weights accordingly. They must be accessible by the Docker container via the file path `/opt/data/weights/`.

### CODE REVIEW

**Comment by lewfish** [comment]
> They must be accessible by the docker via the file path /opt/data/weights/

FYI, people usually refer to a container as a "container" or a "Docker container", but not as a "docker".

**Comment by lewfish** [comment]
> You may optionally connect an AWS bucket and download the weights if they do not already exist locally. In order to do so, the docker must be given valid AWS credentials.

I think we should just copy the file over via the console, or via the command line in the Vagrant box. Giving credentials to the container is more trouble than it's worth and could cause some unintended consequences given the way things are written right now.

File: src/rastervision/common/models/custom_layers.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from keras import initializers
```
Comment (lewfish) [inline src/rastervision/common/models/custom_layers.py]: You should add comments at the top of these three files noting where they were modified from as well as the license of the repo they come from.

Comment (lewfish) [inline src/rastervision/common/models/custom_layers.py]: I see you did it for the other two files ... just might want to move it to the top since the stuff we copied spans more than the function that is commented.

File: src/rastervision/common/models/densenet121.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    x_fc = GlobalAveragePooling2D(name='pool'+str(final_stage))(x)
+    x_fc = Dense(1000, name='fc6')(x_fc)
+    x_fc = Activation('softmax', name='prob')(x_fc)
+
+    model = Model(img_input, x_fc, name='densenet')
+
+    # load weights
+    if weights == 'imagenet':
+        weights_path = '/opt/data/weights/densenet121_weights_tf.h5'
```
Comment (lewfish) [inline src/rastervision/common/models/densenet121.py]: Instead of hardcoding this path, you should do something like `join(weights_path, 'densenet121_weights_tf.h5')`.

File: src/rastervision/common/models/densenet169.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    x_fc = GlobalAveragePooling2D(name='pool'+str(final_stage))(x)
+    x_fc = Dense(1000, name='fc6')(x_fc)
+    x_fc = Activation('softmax', name='prob')(x_fc)
+
+    model = Model(img_input, x_fc, name='densenet')
+
+    # load weights
+    if weights == 'imagenet':
+        weights_path = '/opt/data/weights/densenet169_weights_tf.h5'
```
Comment (lewfish) [inline src/rastervision/common/models/densenet169.py]: Same thing here.

File: src/rastervision/tagging/models/factory.py [hunk 4]
```diff
@@ -1,7 +1,11 @@
 from rastervision.common.models.factory import ModelFactory
 from rastervision.common.models.resnet50 import ResNet50
+from rastervision.common.models.densenet121 import DenseNet121
+from rastervision.common.models.densenet169 import DenseNet169
 
 BASELINE_RESNET = 'baseline_resnet'
+DENSENET_121 = "densenet121"
+DENSENET_169 = "densenet169"
```
Comment (lewfish) [inline src/rastervision/tagging/models/factory.py]: Double quotes should be single quotes.

File: src/rastervision/common/models/densenet121.py [hunk 5]
```diff
@@ -1,4 +1,13 @@
-# -*- coding: utf-8 -*-
+'''
+-*- coding: utf-8 -*-
```
Comment (lewfish) [inline src/rastervision/common/models/densenet121.py]: What does this line mean?

File: src/rastervision/common/models/densenet169.py [hunk 6]
```diff
@@ -85,10 +90,10 @@ def DenseNet169(nb_dense_block=4, growth_rate=32, nb_filter=64, reduction=0.5,
 
     # load weights
     if weights == 'imagenet':
-        weights_path = '/opt/data/weights/densenet169_weights_tf.h5'
-        if not exists(weights_path):
+        path = join(weights_path, 'densenet169_weights_tf.h5')
+        if not exists(path):
             download_weights("densenet169_weights_tf.h5")
```
Comment (lewfish) [inline src/rastervision/common/models/densenet169.py]: Some more double quotes. Maybe you should do a search and replace on these files.

**Comment by lewfish** [comment]
After running `./scripts/test` a whole bunch of lint errors show up. I think we can just suppress the errors for the files you copied and pasted by adding this line `# flake8: noqa` to the top of the file which is done in https://github.com/azavea/raster-vision/blob/develop/src/rastervision/common/models/resnet50.py. But you should fix the errors for the other files.

File: src/rastervision/common/models/densenet121.py (hunk 5 above)
Comment (citerana) [inline src/rastervision/common/models/densenet121.py]: As a comment, it requires that the file be read by a certain python codec. I don't think it does any harm as it's only in the copied files.

File: src/rastervision/common/models/densenet121.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+    model = Model(img_input, x_fc, name='densenet')
+
+    # load weights
+    if weights == 'imagenet':
+        path = join(weights_path, 'densenet121_weights_tf.h5')
+        if not exists(path):
+            download_weights("densenet121_weights_tf.h5")
+        model.load_weights(path, by_name=True)
+
```
Comment (citerana) [inline src/rastervision/common/models/densenet121.py]: fixed - flake8 was crashing on some of the copied files so it had been auto-disabled...

**Comment by lewfish** [comment]
It's easier to review if there is a separate commit added for each round of review.
